### PR TITLE
Implement ResultSet packing

### PR DESCRIPTION
**Pull request by desmondcheongzx**

### CODE REVIEW

**Comment by desmondcheongzx** [comment]
This pr addresses the error in #6. Additionally, it implements the predicate push-up optimisation as described in #5, i.e. we delay filtering data points as far as possible until the number of series to filter is minimised.

Query evaluation now works on the following test case. Say we're given 3 data points:
```
{"Write": {"name": "disk", "labels": {"hostname": "host_desmond"}, "variables": {"total": 10}, "timestamp": "2016-01-01 00:04:00+00:00"}}
{"Write": {"name": "disk", "labels": {"hostname": "host_desmond"}, "variables": {"total": 2}, "timestamp": "2016-01-01 00:05:00+00:00"}}
{"Write": {"name": "disk", "labels": {"hostname": "host_desmond"}, "variables": {"total": 4}, "timestamp": "2016-01-01 00:06:00+00:00"}}
```
And the following select query:
```
{"Select": {"name": "test", "predicate":{"name": "test","condition":{"Leaf":{"lhs": {"LabelKey": "hostname"},"rhs": {"LabelValue": "host_desmond"},"op": "Eq"}}}}}
```
This gives us the correct sorted results:
```
Received statement: Select { name: "test", predicate: Predicate { name: "test", condition: Leaf(Condition { lhs: LabelKey("hostname"), rhs: LabelValue("host_desmond"), op: Eq }) } }
Received result: [Record { name: "disk", labels: {"hostname": "host_desmond"}, variables: {"total": 1099511627776.0}, timestamp: 2016-01-01T00:04:00Z }, Record { name: "disk", labels: {"hostname": "host_desmond"}, variables: {"total": 2.0}, timestamp: 2016-01-01T00:05:00Z }, Record { name: "disk", labels: {"hostname": "host_desmond"}, variables: {"total": 4.0}, timestamp: 2016-01-01T00:06:00Z }]
```
. Additionally, we can evaluate more complex queries such as with an "AND" statement here
```
{"Select": {"name": "test", "predicate":{"name": "test","condition":{"And":[{"Leaf":{"lhs": {"LabelKey": "hostname"},"rhs": {"LabelValue": "host_desmond"},"op": "Eq"}},{"Leaf":{"lhs": {"Variable": "total"},"rhs": {"Metric": 4},"op": "LtEq"}}]}}}}
```
which returns the correct single data point:
```
Received statement: Select { name: "test", predicate: Predicate { name: "test", condition: And(Leaf(Condition { lhs: LabelKey("hostname"), rhs: LabelValue("host_desmond"), op: Eq }), Leaf(Condition { lhs: Variable("total"), rhs: Metric(4.0), op: Lt })) } }
Received result: [Record { name: "disk", labels: {"hostname": "host_desmond"}, variables: {"total": 2.0}, timestamp: 2016-01-01T00:05:00Z }]
```

**Comment by desmondcheongzx** [comment]
@n-young rebased and ready for review!

File: src/server/operators/select.rs [hunk 1]
```diff
@@ -115,34 +118,40 @@ impl Conditions {
...
             }
             // If an or, union the results.
             Conditions::Or(b1, b2) => {
-                let r1 = (*b1).eval(shared_block);
+                let mut r1 = (*b1).eval(shared_block);
                 let r2 = (*b2).eval(shared_block);
-                r1.union(r2)
+                r1.union(r2, shared_block);
+                r1.unpack(shared_block);
+                r1
```
Comment (n-young) [inline src/server/operators/select.rs:140]: Union returns an unpacked, so this unpack shouldn't do anything right?

**Review by n-young — APPROVED** [review]

**Comment by n-young** [comment]
LGTM! Sorry this took so long to review :(

File: src/server/operators/select.rs (hunk 1 above)
Comment (desmondcheongzx) [reply src/server/operators/select.rs:140]: Ah not necessarily: if a union encounters two packed ResultSets, AND both ResultSets don't have filters, then it doesn't unpack them.

**Pull request merged**